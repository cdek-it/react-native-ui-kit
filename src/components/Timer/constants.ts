export const COUNTER_SIZE = 28
