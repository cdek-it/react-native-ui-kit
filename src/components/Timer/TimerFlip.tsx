import React, { memo, useEffect, useState } from 'react'
import { View, Text } from 'react-native'
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  runOnJS,
} from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

import { COUNTER_SIZE } from './constants'

interface TimerFlipProps {
  readonly value: number
  readonly duration?: number
}

export const TimerFlip = memo<TimerFlipProps>(({ value, duration = 300 }) => {
  const styles = useStyles()
  const [currentValue, setCurrentValue] = useState(value)
  const [nextValue, setNextValue] = useState<number | null>(value)
  const progress = useSharedValue(0)

  const currentStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: progress.value * COUNTER_SIZE }],
    opacity: 1 - progress.value,
  }))

  const nextStyle = useAnimatedStyle(() => ({
    transform: [{ translateY: progress.value * COUNTER_SIZE }],
    opacity: progress.value,
    bottom: COUNTER_SIZE,
  }))

  useEffect(() => {
    if (nextValue === currentValue && value !== currentValue) {
      setNextValue(null)
      progress.value = 0
    }
  }, [currentValue, nextValue, progress, value])

  useEffect(() => {
    if (value === currentValue || nextValue !== null) return

    setNextValue(value)
    progress.value = withTiming(1, { duration }, (finished) => {
      if (finished) {
        runOnJS(setCurrentValue)(value)
      }
    })
  }, [value, currentValue, duration, progress, nextValue])

  return (
    <View style={styles.container}>
      {nextValue !== null && (
        <Animated.View style={[styles.textWrapper, nextStyle]}>
          <Text style={styles.text}>{nextValue}</Text>
        </Animated.View>
      )}
      <Animated.View style={[styles.textWrapper, currentStyle]}>
        <Text style={styles.text}>{currentValue}</Text>
      </Animated.View>
    </View>
  )
})

const useStyles = makeStyles(({ typography }) => ({
  container: {
    overflow: 'hidden',
    width: COUNTER_SIZE,
    height: COUNTER_SIZE,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
  },
  textWrapper: {
    height: COUNTER_SIZE,
    width: COUNTER_SIZE,
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
  },
  text: {
    fontSize: typography.Size['text-base'],
    fontWeight: 700,
    includeFontPadding: false,
    verticalAlign: 'middle',
    color: typography.Color.Surface['text-surface-0'],
  },
}))
