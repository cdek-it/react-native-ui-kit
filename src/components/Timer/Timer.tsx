import { memo, useEffect, useState } from 'react'
import { View } from 'react-native'
import Animated, {
  cancelAnimation,
  Easing,
  useAnimatedProps,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'
import Svg, { Circle } from 'react-native-svg'

import { makeStyles } from '../../utils/makeStyles'

import { TimerFlip } from './TimerFlip'
import { COUNTER_SIZE } from './constants'

// eslint-disable-next-line import-x/no-deprecated
const AnimatedCircle = Animated.createAnimatedComponent(Circle)

interface TimerProps {
  /**
   * Время в секундах, которое нужно отсчитать.
   * Min - 1, Max - 99
   */
  countFrom: number
  /** Функция, которая будет вызвана по завершению отсчета */
  onFinish?: () => void
}

const BORDER_WIDTH = 2

/**
 * Таймер для использования внутри ui kit, не экспортируется для внешнего использования
 * @see https://www.figma.com/design/2ZnL6XPKEpxAHvrlbRvnMu/Template-Tailwind-CSS-(DS)?node-id=270-1514
 */
export const Timer = memo<TimerProps>(({ countFrom, onFinish }) => {
  const styles = useStyles()
  const circleAnimation = useSharedValue(0)
  const [currentTimerValue, setCurrentTimerValue] = useState(countFrom)
  const circumferenceRadius = COUNTER_SIZE / 2 - BORDER_WIDTH
  const circumferenceLength = 2 * Math.PI * circumferenceRadius
  const center = COUNTER_SIZE / 2

  const circleAnimatedProps = useAnimatedProps(() => ({
    strokeDashoffset: -circleAnimation.value * circumferenceLength,
  }))

  useEffect(() => {
    setCurrentTimerValue(countFrom)
    circleAnimation.value = 0
    circleAnimation.value = withTiming(1, {
      duration: countFrom * 1000,
      easing: Easing.linear,
    })

    return () => cancelAnimation(circleAnimation)
  }, [circleAnimation, countFrom])

  useEffect(() => {
    let timeoutId: ReturnType<typeof setTimeout>

    if (currentTimerValue > 0) {
      timeoutId = setTimeout(
        () => setCurrentTimerValue((prev) => prev - 1),
        1000
      )
    }

    return () => clearTimeout(timeoutId)
  }, [currentTimerValue, onFinish])

  useEffect(() => {
    if (currentTimerValue === 0) {
      onFinish?.()
    }
  }, [currentTimerValue, onFinish])

  return (
    <View style={styles.container} testID={TestId.Container}>
      <Svg style={styles.svgContainer}>
        <AnimatedCircle
          animatedProps={circleAnimatedProps}
          cx={center}
          cy={center}
          fill='none'
          origin={[center, center]}
          r={circumferenceRadius}
          rotation={-90}
          stroke={styles.circle.color}
          strokeDasharray={circumferenceLength}
          strokeLinecap='round'
          strokeWidth={BORDER_WIDTH}
        />
      </Svg>

      <TimerFlip value={currentTimerValue} />
    </View>
  )
})

const useStyles = makeStyles(({ typography }) => ({
  container: {
    overflow: 'hidden',
    width: COUNTER_SIZE,
    height: COUNTER_SIZE,
    alignItems: 'center',
    justifyContent: 'center',
  },
  svgContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: COUNTER_SIZE,
    height: COUNTER_SIZE,
  },
  circle: { color: typography.Color.Surface['text-surface-0'] },
}))

export enum TestId {
  Container = 'TimerContainer',
}
