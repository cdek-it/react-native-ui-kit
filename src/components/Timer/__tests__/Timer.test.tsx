import { act, render } from '@testing-library/react-native'

import { Timer } from '../Timer'

describe('Timer', () => {
  test('snapshot', () => {
    jest.useFakeTimers()
    const { toJSON } = render(<Timer countFrom={2} />)

    expect(toJSON()).toMatchSnapshot('before start')

    act(() => {
      jest.advanceTimersByTime(1000)
    })
    act(() => {
      jest.advanceTimersByTime(1000)
    })

    expect(toJSON()).toMatchSnapshot('on finish')
  })

  test('should reset timer on changing countFrom prop', () => {
    jest.useFakeTimers()
    const { rerender, toJSON } = render(<Timer countFrom={5} />)

    act(() => {
      jest.advanceTimersByTime(1000)
    })

    expect(toJSON()).toMatchSnapshot('after 1 s')

    rerender(<Timer countFrom={10} />)

    expect(toJSON()).toMatchSnapshot('after reset')
  })

  test('should handle onFinish', () => {
    jest.useFakeTimers()
    const mockedOnFinish = jest.fn()
    render(<Timer countFrom={1} onFinish={mockedOnFinish} />)

    expect(mockedOnFinish).not.toHaveBeenCalled()

    act(() => {
      jest.runAllTimers()
    })

    expect(mockedOnFinish).toHaveBeenCalled()
  })

  test('should NOT call onFinish until the timer expires', () => {
    jest.useFakeTimers()
    const mockedOnFinish = jest.fn()
    const { unmount } = render(
      <Timer countFrom={1} onFinish={mockedOnFinish} />
    )

    expect(mockedOnFinish).not.toHaveBeenCalled()

    unmount()
    jest.runAllTimers()

    expect(mockedOnFinish).not.toHaveBeenCalled()
  })

  test('should clear timeout on unmount', () => {
    jest.useFakeTimers()
    const mockedClearTimeout = jest.spyOn(global, 'clearTimeout')
    const { unmount } = render(<Timer countFrom={1} />)
    unmount()

    expect(mockedClearTimeout).toHaveBeenCalledOnce()
  })
})
