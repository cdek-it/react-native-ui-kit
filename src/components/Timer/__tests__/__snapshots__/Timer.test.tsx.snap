// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should reset timer on changing countFrom prop: after 1 s 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": 28,
      "justifyContent": "center",
      "overflow": "hidden",
      "width": 28,
    }
  }
  testID="TimerContainer"
>
  <RNSVGSvgView
    bbHeight={28}
    bbWidth={28}
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "height": 28,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 28,
        },
        {
          "flex": 0,
          "height": 28,
          "width": 28,
        },
      ]
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGCircle
        cx={14}
        cy={14}
        fill={null}
        matrix={
          [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            28,
          ]
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        r={12}
        stroke={
          {
            "payload": 4281282351,
            "type": 0,
          }
        }
        strokeDasharray={
          [
            75.39822368615503,
            75.39822368615503,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      {
        "alignItems": "center",
        "height": 28,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
        "width": 28,
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "bottom": 28,
            "opacity": 0,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "bottom": 28,
          "height": 28,
          "justifyContent": "center",
          "opacity": 0,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        4
      </Text>
    </View>
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "opacity": 1,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "height": 28,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        5
      </Text>
    </View>
  </View>
</View>
`;

exports[`Timer should reset timer on changing countFrom prop: after reset 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": 28,
      "justifyContent": "center",
      "overflow": "hidden",
      "width": 28,
    }
  }
  testID="TimerContainer"
>
  <RNSVGSvgView
    bbHeight={28}
    bbWidth={28}
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "height": 28,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 28,
        },
        {
          "flex": 0,
          "height": 28,
          "width": 28,
        },
      ]
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGCircle
        cx={14}
        cy={14}
        fill={null}
        matrix={
          [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            28,
          ]
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        r={12}
        stroke={
          {
            "payload": 4281282351,
            "type": 0,
          }
        }
        strokeDasharray={
          [
            75.39822368615503,
            75.39822368615503,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      {
        "alignItems": "center",
        "height": 28,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
        "width": 28,
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "bottom": 28,
            "opacity": 0,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "bottom": 28,
          "height": 28,
          "justifyContent": "center",
          "opacity": 0,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        4
      </Text>
    </View>
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "opacity": 1,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "height": 28,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        5
      </Text>
    </View>
  </View>
</View>
`;

exports[`Timer snapshot: before start 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": 28,
      "justifyContent": "center",
      "overflow": "hidden",
      "width": 28,
    }
  }
  testID="TimerContainer"
>
  <RNSVGSvgView
    bbHeight={28}
    bbWidth={28}
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "height": 28,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 28,
        },
        {
          "flex": 0,
          "height": 28,
          "width": 28,
        },
      ]
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGCircle
        cx={14}
        cy={14}
        fill={null}
        matrix={
          [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            28,
          ]
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        r={12}
        stroke={
          {
            "payload": 4281282351,
            "type": 0,
          }
        }
        strokeDasharray={
          [
            75.39822368615503,
            75.39822368615503,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      {
        "alignItems": "center",
        "height": 28,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
        "width": 28,
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "bottom": 28,
            "opacity": 0,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "bottom": 28,
          "height": 28,
          "justifyContent": "center",
          "opacity": 0,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        2
      </Text>
    </View>
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "opacity": 1,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "height": 28,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        2
      </Text>
    </View>
  </View>
</View>
`;

exports[`Timer snapshot: on finish 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "height": 28,
      "justifyContent": "center",
      "overflow": "hidden",
      "width": 28,
    }
  }
  testID="TimerContainer"
>
  <RNSVGSvgView
    bbHeight={28}
    bbWidth={28}
    focusable={false}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        {
          "height": 28,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 28,
        },
        {
          "flex": 0,
          "height": 28,
          "width": 28,
        },
      ]
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
    >
      <RNSVGCircle
        cx={14}
        cy={14}
        fill={null}
        matrix={
          [
            6.123233995736766e-17,
            -1,
            1,
            6.123233995736766e-17,
            0,
            28,
          ]
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeDasharray",
            "strokeDashoffset",
            "strokeLinecap",
          ]
        }
        r={12}
        stroke={
          {
            "payload": 4281282351,
            "type": 0,
          }
        }
        strokeDasharray={
          [
            75.39822368615503,
            75.39822368615503,
          ]
        }
        strokeDashoffset={null}
        strokeLinecap={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <View
    style={
      {
        "alignItems": "center",
        "height": 28,
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
        "width": 28,
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "bottom": 28,
            "opacity": 0,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "bottom": 28,
          "height": 28,
          "justifyContent": "center",
          "opacity": 0,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        0
      </Text>
    </View>
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "opacity": 1,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          },
        }
      }
      style={
        {
          "alignItems": "center",
          "height": 28,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "transform": [
            {
              "translateY": 0,
            },
          ],
          "width": 28,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2f2f2f",
            "fontSize": 14,
            "fontWeight": 700,
            "includeFontPadding": false,
            "verticalAlign": "middle",
          }
        }
      >
        1
      </Text>
    </View>
  </View>
</View>
`;
